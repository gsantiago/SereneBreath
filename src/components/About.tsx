import React from "react";
import { Popover } from "./Popover";

export function About({ isVisible }) {
  return (
    <Popover isVisible={isVisible} position="bottom">
      <section className="p-3">
        <p>made with love by @gsantiago</p>
      </section>
    </Popover>
  );
}
