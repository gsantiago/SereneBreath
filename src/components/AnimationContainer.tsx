import { animated, SpringValue, useSpringValue } from "@react-spring/web";

export const SIZE = 200;

export interface AnimationContainerProps {
  scale?: SpringValue<number>;
  children: React.ReactNode;
}

export function AnimationContainer({
  scale,
  children,
}: AnimationContainerProps) {
  const defaultScale = useSpringValue(1);

  return (
    <animated.div
      className="flex items-center justify-center "
      style={{ width: SIZE, height: SIZE, scale: scale ?? defaultScale }}
    >
      {children}
    </animated.div>
  );
}
