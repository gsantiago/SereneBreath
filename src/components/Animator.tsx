import React, { useEffect, useState } from "react";
import { AnimationProps } from "@/config/types";

export interface AnimatorProps
  extends Omit<AnimationProps, "state" | "isHolding"> {
  animation: React.FC<AnimationProps>;
}

export function Animator({
  animation: Animation,
  currentStep,
  pattern,
}: AnimatorProps) {
  const [state, setState] = useState<AnimationProps["state"]>("idle");

  const isHolding = currentStep === 1 || currentStep === 3;

  useEffect(() => {
    if (currentStep === 0 || currentStep === 1) {
      setState("inhaling");
    } else {
      setState("exhaling");
    }
  }, [currentStep]);

  return (
    <Animation
      currentStep={currentStep}
      pattern={pattern}
      state={state}
      isHolding={isHolding}
    />
  );
}
