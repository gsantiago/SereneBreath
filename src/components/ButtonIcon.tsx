import React from "react";

export interface ButtonIconProps
  extends Omit<React.HTMLProps<HTMLButtonElement>, "type"> {
  isActive?: boolean;
  position?: "top" | "bottom";
}

export function ButtonIcon({
  isActive,
  position = "top",
  ...props
}: ButtonIconProps) {
  const positionClass =
    position === "top" ? "rounded-t-md" : "rounded-b-md shadow-md";

  return (
    <button
      type="button"
      className={`px-2 pb-2 pt-2 text-3xl dark:text-white ${
        isActive
          ? `z-20 bg-white dark:bg-slate-800 ${positionClass}`
          : "rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
      }`}
      {...props}
    />
  );
}
