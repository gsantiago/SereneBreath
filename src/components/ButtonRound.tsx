import React, { HTMLAttributes } from "react";

export function ButtonRound({
  style,
  ...props
}: HTMLAttributes<HTMLButtonElement>) {
  return (
    <button
      className="flex h-12 w-12 items-center justify-center rounded-full bg-teal-500 text-2xl text-white hover:bg-teal-600"
      {...props}
    />
  );
}
