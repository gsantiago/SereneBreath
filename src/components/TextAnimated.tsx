import React, { useEffect } from "react";
import { animated, useSpring } from "@react-spring/web";

export interface TextAnimatedProps
  extends React.HTMLAttributes<HTMLParagraphElement> {
  children: string;
}

export function TextAnimated({ style, ...props }: TextAnimatedProps) {
  const [springs, api] = useSpring(() => ({
    from: { opacity: 0, y: -10 },
  }));

  useEffect(() => {
    api.start({
      from: {
        opacity: 0,
        y: -10,
      },
      to: {
        opacity: 1,
        y: 0,
      },
    });
  }, [props.children]);

  return <animated.p style={{ ...style, ...springs }} {...props} />;
}
