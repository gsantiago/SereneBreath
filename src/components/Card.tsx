import { RingAnimation } from "@/animations/RingAnimation";
import { Animator } from "@/components/Animator";
import { Pattern } from "@/config/types";

export const WIDTH = 250;
const HEIGHT = "100%";

export interface CardProps {
  title: string;
  description: string;
  pattern: Pattern;
}

export function Card({ title, description, pattern }: CardProps) {
  return (
    <section
      className="overflow-hidden rounded-lg bg-white"
      style={{ width: WIDTH, height: HEIGHT, paddingBottom: 45 }}
    >
      <div className="flex h-1/2 w-full items-center justify-center bg-orange-100">
        <div style={{ transform: "scale(0.8)" }}>
          <Animator
            currentStep={3}
            pattern={[2, 2, 2, 2]}
            animation={RingAnimation}
          />
        </div>
      </div>
      <div className="p-5">
        <h1 className="text-xl font-bold sm:text-2xl">{title}</h1>
        <p className="my-1 text-xs text-gray-400">{pattern.join("-")}</p>
        <p className="text-sm text-gray-800 sm:text-base">{description}</p>
      </div>
    </section>
  );
}
