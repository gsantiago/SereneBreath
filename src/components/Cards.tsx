import { Carousel } from "@/components/Carousel";
import { Card, WIDTH as CARD_WIDTH } from "@/components/Card";

import { useTranslation } from "@/hooks/useTranslation";
import { useStorage } from "@/hooks/useStorage";

import { techniques } from "@/config/techniques";
import { Technique } from "@/config/types";

export function Cards() {
  const { t } = useTranslation();

  const [technique, setTechnique] = useStorage("technique");
  const activeIndex = techniques.findIndex((t) => t.name === technique);

  return (
    <Carousel<Technique>
      activeIndex={activeIndex}
      itemWidth={CARD_WIDTH}
      items={[...techniques]}
      onChange={(item) => setTechnique(item.name)}
    >
      {({ item, isActive }) => (
        <Card
          title={t(`techniques.${item.name}.title`)}
          description={t(`techniques.${item.name}.description`)}
          pattern={item.pattern}
          isActive={isActive}
        />
      )}
    </Carousel>
  );
}
