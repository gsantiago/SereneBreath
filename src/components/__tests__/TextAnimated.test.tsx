import { render, axe, screen } from "@/test-utils";
import { TextAnimated } from "@/components/TextAnimated";

const setup = () => render(<TextAnimated>One</TextAnimated>);

test("renders successfuly", () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();
});

test("accessibility", async () => {
  const { container } = setup();
  expect(await axe(container)).toHaveNoViolations();
});

test("text changing", () => {
  const { rerender } = setup();

  expect(screen.getByText("One")).toBeInTheDocument();

  rerender(<TextAnimated>Two</TextAnimated>);

  expect(screen.getByText("Two")).toBeInTheDocument();
  expect(screen.queryByText("One")).not.toBeInTheDocument();
});
