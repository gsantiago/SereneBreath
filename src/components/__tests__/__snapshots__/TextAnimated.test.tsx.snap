// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders successfuly 1`] = `
<div>
  <p
    style="opacity: 0; transform: translate3d(0,-10px,0);"
  >
    One
  </p>
</div>
`;
